### PR TITLE
Extend the scripts to auto-download the maven-wrapper.jar from maven central

### PR DESCRIPTION
**Pull request by chrisdutz**
I extended the scripts to allow usage of maven-wrapper without requiring the jar to be checked in with the rest of the sources. If the jar is found (because it is checked in or if it had been downloaded peviously) this is used, if not several options are tried out depending on the platform:
- Windows: 
    - powershel is used to download

- Linux/Mac: 
    - curl
    - wget 
    - if both are not available a java class is compiled and executed to do the download.

I didn't add the java option to the windows script for now as far as I know the powershel option should be available on all windows systems.

### CODE REVIEW

**Comment by atanasenko** [comment]
Wrapper download url should also be configurable, same as with maven distro that gets downloaded by the wrapper

File: mvnw [hunk 1]
```diff
@@ -200,6 +200,42 @@ if [ -z "$BASE_DIR" ]; then
...
+##########################################################################################
+# Extension to allow automatically downloading the maven-wrapper.jar from Maven-central
+# This allows using the maven wrapper in projects that prohibit checking in binary data.
+##########################################################################################
+JAR_URL="http://central.maven.org/maven2/io/takari/maven-wrapper/0.2.1/maven-wrapper-0.2.1.jar"
+if [ -r "$BASE_DIR/.mvn/wrapper/maven-wrapper.jar" ]; then
+    echo "Found .mvn/wrapper/maven-wrapper.jar"
+else
+    echo "Couldn't find .mvn/wrapper/maven-wrapper.jar, downloading it ..."
+    if which wget >/dev/null; then
```
Comment (gzm55) [inline mvnw]: use posix way to check cmd: `command -v wget > /dev/null`

**Comment by davidrabinowitz** [comment]
+1

File: mvnw (hunk 1 above)
Comment (chrisdutz) [reply mvnw]: Ok ... just pushed that change (I really would like to have github send me emails on responses like this)

**Comment by chrisdutz** [comment]
Right now I'm a little confused as to what is still missing ... if you guys could sum that up for me, I'd be more than happy to address those things. But keep in mind ... I am a total new-be to this writing of sh scripts ... so please write things as if you were explaining things to a child ;-) ("making url configurable" -(does this translate to)-> "add a property to maven-wrapper.properties containing the url and use that in the script")

**Comment by chrisdutz** [comment]
I just had a look and it seems as if maven-wrapper.jar uses maven-wrapper.properties to find where to download things ... this isn't used by the mvnw script directly ... should we be mixing things here? If yes ... how could I load the properties in there and make them accessible in the script?

File: mvnw [hunk 2]
```diff
@@ -200,6 +200,42 @@ if [ -z "$BASE_DIR" ]; then
...
+# This allows using the maven wrapper in projects that prohibit checking in binary data.
+##########################################################################################
+JAR_URL="http://central.maven.org/maven2/io/takari/maven-wrapper/0.2.1/maven-wrapper-0.2.1.jar"
+if [ -r "$BASE_DIR/.mvn/wrapper/maven-wrapper.jar" ]; then
+    echo "Found .mvn/wrapper/maven-wrapper.jar"
+else
+    echo "Couldn't find .mvn/wrapper/maven-wrapper.jar, downloading it ..."
+    if command -v wget > /dev/null; then
+        echo "Found wget ... using wget"
+        wget ‐‐output-document=.mvn/wrapper/maven-wrapper.jar "$JAR_URL"
```
Comment (gzm55) [inline mvnw]: in some environment, such as in alpine, wget has no ‐‐output-document option, only '-O'  works.

**Comment by gzm55** [comment]
For a configurable url, we can add a optional key (e.g. wrapperUrl) in maven-wrapper.properties, and parse it manually first before downloading the wrapper jar.

For linux, we can parse properties file like
```sh
jarUrl='a-default-url'
while IFS="=" read key value; do
  case "$key" in
  (wrapperUrl) jarUrl="$value"; break ;;
  esac
done < "path/to/maven-wrapper.properties"
```

For windows, the same logic works, but need to translate to for-statement in cmd script of windows.

File: mvnw (hunk 2 above)
Comment (chrisdutz) [reply mvnw]: Does "-O" work everywhere?

Comment (gzm55) [reply mvnw]: almost, at least better than ‐‐output-document

Comment (chrisdutz) [reply mvnw]: Ok ... so I updated the pull request.

**Comment by chrisdutz** [comment]
So the last change you would like to have me do, would be to make the URL configurable? This Tree-Style-Comments are making it really hard to follow the conversation.

**Comment by chrisdutz** [comment]
Ok ... so now the downloadUrl has a default and is overridable in the properties file ... anything else missing?

File: mvnw [hunk 3]
```diff
@@ -200,6 +200,49 @@ if [ -z "$BASE_DIR" ]; then
...
+    while IFS="=" read key value; do
+      case "$key" in (wrapperUrl) jarUrl="$value"; break ;;
+      esac
+    done < "$BASE_DIR/.mvn/wrapper/maven-wrapper.properties"
+    echo "Downloading from: $jarUrl"
+
+    if command -v wgety > /dev/null; then
+        echo "Found wget ... using wget"
+        wget ‐O "$BASE_DIR/.mvn/wrapper/maven-wrapper.jar" "$jarUrl"
+    elif command -v curly > /dev/null; then
```
Comment (gzm55) [inline mvnw]: `wgety` and `curly` seems typo of `wget` and `curl`, is it?

**Comment by chrisdutz** [comment]
Again ... anything else missing?

**Comment by mosabua** [comment]
Mostly time to review and test. Before we merge I need to do another feature request (removing output and making it configurable) and this would need to be tested on linux, osx, window and solaris

**Comment by mosabua** [comment]
So I cut a new release without this. I am not sure how we can get this tested on the current supported shell and OS versions to be honest.

File: .mvn/wrapper/maven-wrapper.properties [hunk 4]
```diff
@@ -1 +1,2 @@
-distributionUrl=https://repo1.maven.org/maven2/org/apache/maven/apache-maven/3.5.0/apache-maven-3.5.0-bin.zip
\ No newline at end of file
+distributionUrl=https://repo1.maven.org/maven2/org/apache/maven/apache-maven/3.5.0/apache-maven-3.5.0-bin.zip
+#wrapperUrl=http://central.maven.org/maven2/io/takari/maven-wrapper/0.2.1/maven-wrapper-0.2.1.jar
```
Comment (mosabua) [inline .mvn/wrapper/maven-wrapper.properties]: Should be the https URL and also be the same repo1.maven.org as above

Comment (mosabua) [reply .mvn/wrapper/maven-wrapper.properties]: Also should this really be commented out?

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+import java.nio.channels.*;
+import java.util.Properties;
+
+public class MavenWrapperDownloader {
+
+    /**
+     * Default URL to download the maven-wrapper.jar from, if no 'downloadUrl' is provided.
+     */
+    private static final String DEFAULT_DOWNLOAD_URL =
+            "http://central.maven.org/maven2/io/takari/maven-wrapper/0.2.1/maven-wrapper-0.2.1.jar";
```
Comment (mosabua) [inline .mvn/wrapper/MavenWrapperDownloader.java]: Should be HTTPS and repo1.maven.org... and maybe have version set somehow more cleanly.

File: .mvn/wrapper/maven-wrapper.properties (hunk 4 above)
Comment (chrisdutz) [reply .mvn/wrapper/maven-wrapper.properties]: Well it is in there as an example of how to configure the wrapper. It is equal to the default used in the wrapper and downloader code. Sort of wanted to imply that you don't need that property, but you could use it and how you would have to use it.

File: .mvn/wrapper/MavenWrapperDownloader.java (hunk 5 above)
Comment (chrisdutz) [reply .mvn/wrapper/MavenWrapperDownloader.java]: I did change the urls to the https versions.

**Comment by chrisdutz** [comment]
Ok so I adressed the https issue and just comitted my changes. One way to test this on a variety of systems would be for me to use the updated wrapper in one of the Apache Projects I'm working on. There I could run builds on Windows, Mac, multiple Linux distributions. I did manually test it on Mac, Linux and Windows.

**Comment by chrisdutz** [comment]
Also just updated the version of the jar to the latest version. It would be easy to inject the current version, if the mvnw and mvnw.cmd etc. would be inside the sources so the maven resource plugin could do the filtering and injecting of properties.

**Comment by mosabua** [comment]
Replaced by https://github.com/takari/maven-wrapper/pull/70
